File: baselibrary/src/main/java/com/minicart/android/baselibrary/dagger2/delegate/FragmentDelegateImpl.java
package com.minicart.android.baselibrary.dagger2.delegate;

import android.content.Context;
import android.os.Bundle;
import android.os.Parcel;
import android.os.Parcelable;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.view.View;

import com.minicart.android.baselibrary.dagger2.App;

import butterknife.ButterKnife;
import butterknife.Unbinder;

/**
 * Created by jess on 29/04/2017 16:12
 * Contact with jess.yan.effort@gmail.com
 */

public class FragmentDelegateImpl implements FragmentDelegate {
    private static final String TAG = "FragmentDelegateImpl";
    private FragmentManager mFragmentManager;
    private Fragment mFragment;
    private IFragment iFragment;
    private Unbinder mUnbinder;


    public FragmentDelegateImpl(FragmentManager fragmentManager, Fragment fragment) {
        this.mFragmentManager = fragmentManager;
        this.mFragment = fragment;
        this.iFragment = (IFragment) fragment;
    }

    @Override
    public void onAttach(Context context) {

    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        iFragment.setupFragmentComponent(((App) mFragment.getActivity().getApplication()).getAppComponent());
    }

    @Override
    public void onCreateView(View view, Bundle savedInstanceState) {
        //绑定到butterknife
        if (view != null) {
            mUnbinder = ButterKnife.bind(mFragment, view);
        }
    }

    @Override
    public void onActivityCreate(Bundle savedInstanceState) {
        iFragment.initView(savedInstanceState);
        iFragment.initData(savedInstanceState);
        iFragment.initListener(savedInstanceState);
    }

    @Override
    public void onStart() {

    }

    @Override
    public void onResume() {

    }

    @Override
    public void onPause() {

    }

    @Override
    public void onStop() {

    }

    @Override
    public void onSaveInstanceState(Bundle outState) {

    }

    @Override
    public void onDestroyView() {
        if (mUnbinder != null && mUnbinder != Unbinder.EMPTY) {
            try {
                mUnbinder.unbind();
            } catch (IllegalStateException e) {
                e.printStackTrace();
                //fix Bindings already cleared
            }
        }
    }

    @Override
    public void onDestroy() {
        this.mUnbinder = null;
        this.mFragmentManager = null;
        this.mFragment = null;
        this.iFragment = null;
    }

    @Override
    public void onDetach() {

    }

    /**
     * Return true if the fragment is currently added to its activity.
     */
    @Override
    public boolean isAdded() {
        return mFragment == null ? false : mFragment.isAdded();
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {

    }

    protected FragmentDelegateImpl(Parcel in) {
        this.mFragmentManager = in.readParcelable(FragmentManager.class.getClassLoader());
        this.mFragment = in.readParcelable(Fragment.class.getClassLoader());
        this.iFragment = in.readParcelable(IFragment.class.getClassLoader());
        this.mUnbinder = in.readParcelable(Unbinder.class.getClassLoader());
    }

    public static final Parcelable.Creator<FragmentDelegateImpl> CREATOR = new Parcelable.Creator<FragmentDelegateImpl>() {
        @Override
        public FragmentDelegateImpl createFromParcel(Parcel source) {
            return new FragmentDelegateImpl(source);
        }

        @Override
        public FragmentDelegateImpl[] newArray(int size) {
            return new FragmentDelegateImpl[size];
        }
    };
}

File: baselibrary-kotlin/src/main/java/com/minicart/android/baselibrary/dagger2/delegate/FragmentDelegateImpl.kt
package com.minicart.android.baselibrary.dagger2.delegate

import android.content.Context
import android.os.Bundle
import android.os.Parcel
import android.os.Parcelable
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager
import android.view.View
import butterknife.ButterKnife
import butterknife.Unbinder
import com.minicart.android.baselibrary.dagger2.App

/**
 * Created by jess on 29/04/2017 16:12
 * Contact with jess.yan.effort@gmail.com
 */

class FragmentDelegateImpl : FragmentDelegate {
    private var mFragmentManager: FragmentManager? = null
    private var mFragment: Fragment? = null
    private var iFragment: IFragment? = null
    private var mUnbinder: Unbinder? = null


    constructor(fragmentManager: FragmentManager, fragment: Fragment) {
        this.mFragmentManager = fragmentManager
        this.mFragment = fragment
        this.iFragment = fragment as IFragment
    }

    override fun onAttach(context: Context) {

    }

    override fun onCreate(savedInstanceState: Bundle) {
        iFragment!!.setupFragmentComponent((mFragment!!.activity.application as App).appComponent)
    }

    override fun onCreateView(view: View, savedInstanceState: Bundle) {
        //绑定到butterknife
        if (view != null) {
            mUnbinder = ButterKnife.bind(mFragment!!, view)
        }
    }

    override fun onActivityCreate(savedInstanceState: Bundle) {
        iFragment!!.initView(savedInstanceState)
        iFragment!!.initData(savedInstanceState)
        iFragment!!.initListener(savedInstanceState)
    }

    override fun onStart() {

    }

    override fun onResume() {

    }

    override fun onPause() {

    }

    override fun onStop() {

    }

    override fun onSaveInstanceState(outState: Bundle) {

    }

    override fun onDestroyView() {
        if (mUnbinder != null && mUnbinder !== Unbinder.EMPTY) {
            try {
                mUnbinder!!.unbind()
            } catch (e: IllegalStateException) {
                e.printStackTrace()
                //fix Bindings already cleared
            }

        }
    }

    override fun onDestroy() {
        this.mUnbinder = null
        this.mFragmentManager = null
        this.mFragment = null
        this.iFragment = null
    }

    override fun onDetach() {

    }

    /**
     * Return true if the fragment is currently added to its activity.
     */
    override val isAdded: Boolean
        get() = if (mFragment == null) false else mFragment!!.isAdded

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {

    }

    protected constructor(parcel: Parcel) {
        this.mFragmentManager = parcel.readParcelable<Parcelable>(FragmentManager::class.java.classLoader) as FragmentManager
        this.mFragment = parcel.readParcelable<Parcelable>(Fragment::class.java.classLoader) as Fragment
        this.iFragment = parcel.readParcelable<Parcelable>(IFragment::class.java.classLoader) as IFragment
        this.mUnbinder = parcel.readParcelable<Parcelable>(Unbinder::class.java.classLoader) as Unbinder
    }

    companion object {
        private val TAG = "FragmentDelegateImpl"

        val CREATOR: Parcelable.Creator<FragmentDelegateImpl> = object : Parcelable.Creator<FragmentDelegateImpl> {
            override fun createFromParcel(source: Parcel): FragmentDelegateImpl {
                return FragmentDelegateImpl(source)
            }

            override fun newArray(size: Int): Array<FragmentDelegateImpl?> {
                return arrayOfNulls(size)
            }
        }
    }
}
